### ISSUE TITLE
CIF v3 inputs/outputs

### ISSUE THREAD

**Issue by newconf** [csirtgadgets/bearded-avenger#254]
Hi, 
Please can you enlighten me on the input formats supported by CIF (STIX, CyBOX) ? I am a student and have to do a survey on different threat intelligence tools, but I need some more information to do it correctly.
Are you OK with the last line of this table? Thank you in advance.

<img width="520" alt="compare" src="https://cloud.githubusercontent.com/assets/25781820/23125088/693b9584-f770-11e6-850b-0ff3a5af4691.png">

**Comment by wesyoung** [comment]
hiya,

we're not quite to the point where v3 is completely done and documented, but it looks like you have a good start. if you wanna take a look at the v2 doc, it might give you a better sense:

https://github.com/csirtgadgets/massive-octo-spice/wiki/The-CIF-Book

also if you want to get your hands dirty; you can mess around with:

https://github.com/csirtgadgets/csirtg-smrt-py

more specifically:

https://github.com/csirtgadgets/csirtg-smrt-py/tree/master/csirtg_smrt/parser

you can also find a number of papers; books, blogposts [older] about v1 and v2 that may help answer more of your questions (and or just standing up a VM and messing around with the data).

**Comment by newconf** [comment]
Thank you very much for your answer and the useful links!
So I understand that STIX is partially parsed through: https://github.com/csirtgadgets/csirtg-smrt-py/tree/master/csirtg_smrt/parser
And for exchanging data between peers, are STIX/TAXII supported ? otherwise, how to do it? 
Thks

**Comment by wesyoung** [comment]
csirtg-smrt will parse *very basic* STIX data into our basic csirtg-indicator-py format (very rarely used, but it's there to help STIX users wrap their head around csirtg-smrt). the cif-sdk will also dump csirtg-indicator-py data into a very basic stix format:

https://github.com/csirtgadgets/csirtg-indicator-py/blob/master/csirtg_indicator/format/zstix.py

but we do not support any TAXII stuff at the moment because it's based on legacy technology thats too slow for what we need it to do (same reason we don't really support much STIX either).

there's been some discussion about adding in a TAXII transport listener to csirtg-smrt:

https://groups.google.com/forum/#!searchin/ci-framework/taxii%7Csort:relevance

but we haven't seen a PR yet from anyone and it hasn't made the top of my list. CIF has an extremely high speed requirement and is heavily based on ZeroMQ for that reason. we enable users to use some REST stuff when interacting with it; but more and more of the interface is being developed around ZeroMQ to handle the amount of data traversing the architecture.

 i usually inform users; that if STIX/TAXII are a heavy requirement for what you're trying to do (ie: work in the gov, etc), then CIF is probably not the best tool for you and you should seek something else out. we really have very little intention of supporting those technologies moving forward because they're too slow and poorly thought out (imo, take that with a grain of salt of course).

exchanging data between peers is pretty easy with CIFv2 and v3, where-as you hand an API token to your peer and use the smrt plugins to pull data with normal REST/JSON calls. in v3, we're adding in a newer ZeroMQ technology called "ZYRE":

https://github.com/zeromq/zyre
https://github.com/wesyoung/pyzyre

which will enable users to connect their cif-routers together over a p2p style framework (we're working on this now, testing it with live darknet data). we haven't added this into the main code base yet, but we've been working on prototyping it with live data to see where things fall over first.

we also recently wrote a "cif-tail" tool, that helps users "tail" each-others instances (assuming they have the correct API token):

https://github.com/csirtgadgets/bearded-avenger-sdk-py/pull/42

which allows you to setup filters for what kind of data you'd like to pull from the instance. so the routers are being setup to share data in realtime, but with the added capacity of polling data from the instances with more refined filters in not-so-realtime. in most cases, our end users prefer this method over the realtime method because it gives them some control over only pulling "more highly confident" data than "drinking from the firehose". moving forward though, we have some users who are waiting for the firehose so they can do their own style of correlation around the data, we've just been working to standardize the process a bit and zoom in on a protocol that can handle our speed/efficiency requirement (ie: csirtg-indicator-py).

does that make sense? :)

**Comment by newconf** [comment]
Very interesting! I will examine deeply the links you shared and add your inputs in my table so that it becomes up to date :)
Regarding the exchange of data, is it true that until now, CIF data could be shared among different CIF instances (in a community) via the "Federation Service" as mentioned in this report (see the link below, Page 6)?
https://www.giac.org/paper/gcih/13998/automated-defense-threat-intelligence-augment/121748
Is it a different thing or does this no longer match the ZeroMQ and newly, the ZYRE technology?
In short, If I should present the exchanging mode of CIF v3 now, what can I say? Can we talk about a community Cloud? thanks!

**Comment by wesyoung** [comment]
in the past; we've taken the HTTP/REST approach, where you could subscribe CIF instances to each-other (just like every other REST based application). that way when you pull data, you're filtering on data you find interesting, not drinking from the firehose.

in v3; the same approach applies, but we're building some hooks for ZYRE (p2p) so you can start linking up firehoses between each-other. no central broker (that's now how the internet scales), but the ability to link up real-time exchanges with your peers as you see fit (we're not getting in the way).

in the higher ed community (REN-ISAC), we run a centralized service for institutions, but it's more of an ease-of-use for orgs that don't wanna run their own infra. many larger orgs run their own infra, tie to us, and then to other partnerships they have [that we may not].

make sense?
